torch.compile + selective activation checkpointing

yf225 opened the pull request:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/105489

NOTE: this PR is tagged "not user facing", because it's not ready to be announced externally yet.

This PR implements torch.compile + selective activation checkpoint (SAC) integration, by using `TagActivationCheckpoint` (same backend as torch.compile + full activation checkpoint integration).

TorchDispatchMode based implementation cannot support including inplace ops in the checkpointed region at the moment (the reason for this needs investigation), and there is also no way to ban them (because TorchDispatchMode now only sees "after-functionalization" ops, so can't detect if an op is in-place). Hence we hide torch.compile + SAC behind a flag (`torch._dynamo.config._experimental_support_context_fn_in_torch_utils_checkpoint`) and will only use it internally for cases that are known to not have in-place ops. This state won't last too long, because in-place op will at least be able to be detected after Brian's mode reordering and related functionalization changes.
So next steps after this PR:
1. Wait for Brian's mode reordering and related functionalization changes to land, and then try to enable the "inplace ops" unit test for torch.compile + selective activation checkpoint (if it doesn't work, investigate why).
2. Unify selective- and full-checkpoint under TorchDispatchMode based implementation.

Differential Revision: D47497145



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng @anijain2305 @ipiszy

anijain2305 commented on torch/_higher_order_ops/wrap.py:
```diff
@@ -110,24 +442,60 @@ def divide_kwargs(kwargs):
+            # Using interpreter allows preservation of metadata through torch.compile stack.
+            with fx_traceback.preserve_node_meta():
+                return checkpoint(Interpreter(gmod).run, *args, **kwargs)
+        else:
```
IMO, having a switch to explicitly go one way or the other is a better approach. This is with the assumption that SAC is a niche usecase.

AOT Autograd traced checkpoint requires more work around CSE etc. As that part of the stack is not hardened. I think its better if we have well tested path as default. And when that path does not work out, we emit an actionable error message to switch to the less tested path.

yf225 replied:
CSE seems to not cause problem with selective checkpointing right now actually, but yes I can make this into a flag.

anijain2305 replied:
Have you checked the memory numbers after torch.compile? CSE will silently undo AC, and will increase memory footprint.

yf225 replied:
Note to self: check memory usage of SAC under torch.compile

Also found that `min_cut_rematerialization_partition` (which has CSE) is not used when SAC is applied, and it might explain why we don't see CSE causing problem with SAC.

yf225 replied:
Update: I found a way to use `node.meta["recompute"]` for selective checkpointing case that's similar to full-checkpointing, details are in `CachedTorchDispatchMode._handle_compile_in_forward_ctx` method

Chillee commented:
I don't think we actually need to ban in-place ops right? Unless they modify one of the inputs?

yf225 commented on test/dynamo/test_activation_checkpointing.py:
```diff
@@ -324,13 +391,292 @@ def fn(x, y, z):
+            AssertionError,
+            "In-place ops are not supported in selective checkpointing region under torch.compile"
+        ):
+            self._validate(fn, backend, x, y)
```
@Chillee I compared this test case with `test_tags_selective_checkpoint_outplace_op` above (`torch.selu_` vs. `torch.selu`), and it seems that as long as we have in-place op applied to inputs or activations in the graph, the gradient comparison check would fail.

Chillee replied:
Hmm... I need to think a bit more about this. I would ... expect that functionalization would apply underneath selective checkpointing, and so it would still work correctly.

bdhirsh replied:
Hmm. This seems pretty similar to the "what should run first, torch_dispatch vs. functionalization" ordering issue I'm running into with subclass support. I don't think this will work today, since the "default" thing that happens is that functionalization has a higher priority in the dispatcher than __toch_dispatch__.

I updated the design doc after running into some issues with DTensor - I think functionalization will need to run below torch_dispatch (updated [here](https://docs.google.com/document/d/1COQ291nOZvtFnzGTQMJqoYZ3sttEYFw_7HbfSyL8gcA/edit#bookmark=id.4tvs5sten7fb)). Hopefully that should fix this problem too - I can give a shout once I have a working PR to test with if that sounds good to you all.

bdhirsh replied:
But (once the functionalization ordering issue is sorted out), agreed with Horace's comment - instead of manually checking for inplace ops, would it make sense to use the existing `speculate_graph` infra that Voz/Richard wrote, that just tries to ban subgraphs containing input mutations?

yf225 replied:
yes - I can confirm that with Brian's mode reordering stack, I can use `func._schema.is_mutable` to check inplace ops.

bdhirsh replied:
Hmm @yf225 do we want to ban **all** mutable ops for now? Or is that needed any time soon.

Functionalization should handle any mutable ops inside the checkpointed region. The main issue is just "what if a mutable op tries to mutate one of the **inputs** to the checkpointed region" - and in that case, higher order ops today are supposed to error in `speculate_subgraph` if they detect that there were any input mutations.

yf225 replied:
Update: will wait for Brian's stack to land for enabling in-place op support

albanD commented on torch/_higher_order_ops/wrap.py:
```diff
@@ -25,45 +30,371 @@ def wrapper():
 wrap = Wrap()
 
-class WrapActivationCheckpoint(HigherOrderOperator):
+random_ops = set([
```
Can we find a way not to hardcode these?

bdhirsh replied:
As an alternative (just as an idea, not sure that this is the best approach):

Unfortunately, we don't have a systematic way to get all random ops from the `torch.*` module. But if instead you could get at the ATen ops that will be captured in our checkpointed function, then you can inspect the `tags` on each the `OpOverload` for a `nondeterminism` tag (indicated randomness, [code](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/tags.yaml#L25))

If we have already used dynamo to confirm that the function is traceable, then one option would be to `make_fx()` it and get out a graph of the aten ops.

yf225 replied:
Update: will use Brian's mode reordering stack, and check `.is_mutable` for inplace ops and `nondeterministic_seeded` tag for random ops

albanD commented on torch/_higher_order_ops/wrap.py:
```diff
@@ -25,45 +30,371 @@ def wrapper():
+    "uniform_",
+])
+
+inplace_ops = set([
```
`if name[-1] == "_" or (name.starts_with("__i") and name.ends_with("__")):`

Chillee commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -890,10 +889,14 @@ class CheckpointHigherOrderVariable(WrapHigherOrderVariable):
         from .builder import wrap_fx_proxy
 
-        checkpoint_kwargs, gmod_kwargs = TagActivationCheckpoint.divide_kwargs(kwargs)
+        if "context_fn" in kwargs:
```
So this works with arbitrary context_fns?

yf225 replied:
Only guarantees to work with TorchDispatch based context_fns actually. Maybe we can assert that user must use TorchDispatch functions?

yf225 replied:
Added check in torch/utils/checkpoint.py and corresponding unit test.

yf225 commented on test/dynamo/test_activation_checkpointing.py:
```diff
@@ -38,6 +50,53 @@ def op_count(gm):
+            kwargs = {}
+        if self.policy_fn(func, *args, **kwargs):
+            out = func(*args, **kwargs)
+            self.storage.append(out)
```
Note to self: updated to use detach similar to https://fburl.com/code/oaxjq5n6

yf225 commented on test/dynamo/test_activation_checkpointing.py:
```diff
@@ -324,13 +391,292 @@ def fn(x, y, z):
+                    # Saves output of all compute ops, except second sigmoid
+                    # (i.e. only recomputes second sigmoid in backward pass)
+                    out = func(*args, **kwargs)
+                    self.storage.append(out)
```
Note to self: updated to use detach similar to https://fburl.com/code/oaxjq5n6

yf225 commented on test/dynamo/test_activation_checkpointing.py:
```diff
@@ -361,6 +428,285 @@ def fn(x, y):
+        bw_compiler = functools.partial(
+            count_ops_list,
+            freqs=[4, 0],
+            ops=[torch.ops.aten.mm.default, torch.ops.aten.sigmoid.default],
```
Note to self: check that we are recomputing selu

yf225 commented:
`fx_traceback.current_meta` is another way to track the recompute tag for user dispatch mode

yf225 commented on torch/utils/checkpoint.py:
```diff
@@ -1098,6 +1109,138 @@ def unpack_hook_with_error_cb(holder):
+                out = self.pop_from_storage(func, args, kwargs)
+            else:
+                out = func(*args, **kwargs)
+            return out
```
Throw error (and add unit test) if user's own TDM doesn't handle compile properly

yf225 commented on torch/utils/checkpoint.py:
```diff
@@ -1098,6 +1109,138 @@ def unpack_hook_with_error_cb(holder):
+class CachedTorchDispatchMode(TorchDispatchMode):
+    r"""
+    A :class:`TorchDispatchMode` to implement selective activation checkpointing
+    that's compatible with torch.compile. Used together with CachingTorchDispatchMode.
```
After discussion: we should make Caching and Cached private APIs, and only expose `policy_fn` API to users.

yf225 commented on torch/utils/checkpoint.py:
```diff
@@ -1098,6 +1109,138 @@ def unpack_hook_with_error_cb(holder):
+    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
+        if kwargs is None:
+            kwargs = {}
+        should_not_recompute = self.policy_fn(self, func, *args, **kwargs)
```
we should pass "forward" instead of "self" here

yf225 commented on torch/utils/checkpoint.py:
```diff
@@ -1098,6 +1109,138 @@ def unpack_hook_with_error_cb(holder):
+    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
+        if kwargs is None:
+            kwargs = {}
+        should_not_recompute = self.policy_fn(self, func, *args, **kwargs)
```
we should pass "recompute" instead of "self" here

yf225 commented on test/dynamo/test_activation_checkpointing.py:
```diff
@@ -358,9 +420,291 @@ def fn(x, y):
+            ]
+
+            def _custom_policy(dispatch_mode, func, *args, **kwargs):
+                mm_count_key = f"id{dispatch_mode}_mm_count"
```
```suggestion
                mm_count_key = f"{dispatch_mode}_mm_count"
```

yf225 commented:
@anijain2305 @Chillee I changed the approach so that selective checkpointing also uses the partitioner-based “duplicate late” approach, similar to full-checkpointing. Checked with Horace and he seems okay with it, and he said as long as we can later change back to “duplicate early” when things are ready. To support that vision, I only expose the `policy_fn` (func -> whether to checkpoint) as the public API to users, which is the API that will never change regardless of internal implementation.

anijain2305 commented on test/dynamo/test_activation_checkpointing.py:
```diff
@@ -358,9 +420,291 @@ def fn(x, y):
+            freq=4,
+            op=torch.ops.aten.mm.default,
+        )
+        with _functorch_partitioner_use_cse():
```
You can annotate the test with `@torch._functorch.config.patch(cse=True)`

anijain2305 commented on torch/_higher_order_ops/wrap.py:
```diff
@@ -25,45 +29,34 @@ def wrapper():
     to AOT Autograd, which is ok with tracing saved tensor hooks. As a result of
-    AOT tracing torch.utils.checkpoint code, we have a backward graph with
-    recomputed forward nodes.
+    AOT tracing torch.utils.checkpoint code, particularly _CachingTorchDispatchMode
+    and _CachedTorchDispatchMode, we obtain FX graph with added "recomputable" tag
+    to the nodes that should be recomputed. Then, similar to TagActivationCheckpoint
+    case, we rely on the partitioners to actually duplicate the nodes.
```
This is confusing. Why do we even need `recomputable` tags if we are relying on AOT Autograd to do the tracing.

Imo, we should keep WrapActivationCheckpoint simple - let AOT Autograd take care of everything. In fact, this TorchDispatch facility seems to sit well with TagActivationCheckpoint to me.

anijain2305 commented on torch/_higher_order_ops/wrap.py:
```diff
@@ -110,24 +104,41 @@ def divide_kwargs(kwargs):
+                kwargs["context_fn"] = self.context_fn
+                # We first tag all nodes as "recompute" in this graph, and then we undo the "recompute" tag
+                # for specific nodes in _CachedTorchDispatchMode.
+                gmod = self.tag_nodes(gmod)
```
Yeah, this on face value seems very confusing. WrapActivationCheckpoint (and this is a bad name - my mistake) should not need any tagging. This should go in TagActivationCheckpoint.

anijain2305 commented on torch/utils/checkpoint.py:
```diff
@@ -408,6 +413,10 @@ def checkpoint(
             This argument is only supported if ``use_reentrant=False``.
+            This is useful e.g. for implementing selective checkpointing.
+            When selective checkpointing is used with torch.compile,
+            random ops (e.g. ``torch.dropout``) and in-place ops (e.g. ``torch.relu_``)
```
Just interested - why dont we allow in place ops with SAC?

yf225 replied:
Currently running into gradient mismatch for the inplace test case we have, will wait for Brian's stack to land and then investigate this again

anijain2305 commented on torch/utils/checkpoint.py:
```diff
@@ -1098,6 +1107,166 @@ def unpack_hook_with_error_cb(holder):
+            "In-place ops are not supported in selective checkpointing region under torch.compile. "
+            + f"Found in-place ops: {func}."
+        )
+        assert torch.Tag.nondeterministic_seeded not in func.tags, (
+            "Random ops are not supported in selective checkpointing region under torch.compile. "
+            + f"Found random ops: {func}."
+        )
```
Oh is that always the case? Can we never have random ops in selective checkpointing? Why is that the case?

The whole point of Duplicate Late is to handle rng ops. If no rng ops, Duplicate early + disabling cse might be good enough. So, this complexity of TorchDispatchMode and still using AOT Autograd for tracing seems unnecessary. I am still behind the idea of unifying but currently the stack seems little jarring.

yf225 replied:
Removed restriction on random ops

yf225 commented on torch/utils/checkpoint.py:
```diff
@@ -1098,6 +1107,166 @@ def unpack_hook_with_error_cb(holder):
+        assert not func._schema.is_mutable, (
+            "In-place ops are not supported in selective checkpointing region under torch.compile. "
+            + f"Found in-place ops: {func}."
+        )
```
see if we can remove this inplace restriction

yf225 replied:
we need Brian's stack to make this work. We should add a config to enable this only for experimental use cases.

yf225 commented on torch/utils/checkpoint.py:
```diff
@@ -1098,6 +1107,166 @@ def unpack_hook_with_error_cb(holder):
+        assert torch.Tag.nondeterministic_seeded not in func.tags, (
+            "Random ops are not supported in selective checkpointing region under torch.compile. "
+            + f"Found random ops: {func}."
+        )
```
see if we can remove this random op restriction

yf225 commented on torch/_higher_order_ops/wrap.py:
```diff
@@ -110,24 +104,41 @@ def divide_kwargs(kwargs):
+            gmod = self.tag_nodes(gmod)
+            # Using interpreter allows preservation of metadata through torch.compile stack.
+            with fx_traceback.preserve_node_meta():
+                return Interpreter(gmod).run(*args)
```
we should put SAC+compile logic into TagActivationCheckpoint (and leave the old WrapActivationCheckpoint mode as-is).

Also try to see if full-checkpoint can use the same mode-based `return checkpoint(Interpreter(gmod).run, *args, **kwargs)` logic (by having a special pair of modes for full case).

yf225 replied:
Also try to see if full-checkpoint can use the same mode-based `return checkpoint(Interpreter(gmod).run, *args, **kwargs)` logic (by having a special pair of modes for full case).

anijain2305 commented on torch/_higher_order_ops/wrap.py:
```diff
@@ -110,24 +109,37 @@ def divide_kwargs(kwargs):
+            gmod = self.tag_nodes(gmod)
+            # Using interpreter allows preservation of metadata through torch.compile stack.
+            with fx_traceback.preserve_node_meta():
+                return Interpreter(gmod).run(*args)
```
Have we tried making this path also go through the `checkpoint`? If not too hard, we should make it happen. Maybe we can have dummy context_fn if there is no user passed context_fn

anijain2305 commented on torch/utils/checkpoint.py:
```diff
@@ -1098,6 +1103,158 @@ def unpack_hook_with_error_cb(holder):
+
+def context_fn_gen(policy_fn):
+    """
+    A helper function to generate a pair of contexts to be later passed into
```
It seems like we need this only for torch.compile (even though it can work w/o torch.compile as well). Maybe mention the reason of introducing this wrt torch.compile.

anijain2305 commented on torch/_higher_order_ops/wrap.py:
```diff
@@ -64,17 +69,11 @@ class TagActivationCheckpoint(HigherOrderOperator):
     accepts a Fx graph module which needs to be checkpointed. This operator adds
     "recomputable" tag to the nodes of the Fx graph that should be recomputed.
-
-    The goal is to avoid both Dynamo and AOT Autograd to trace through saved
-    tensor hooks, and rather rely on the partitioners to actually duplicate the
-    nodes. This sits well in the torch.compile stack, because by the time graph
-    reaches partitioner, inductor has already run its functionalization of rng
-    ops. Therefore, the duplication of nodes, by design, respects the rng states
-    in the forward and recomputed forward in backward.
```
I think this is still partially true.

Dynamo still does not trace through saved tensor hooks.

AOT Autograd does trace through saved tensor hooks but has special logic with TorchDispatchMode to override the usual saved_tenso_hooks fn to tag the nodes.

The comment about RNG is still valid.

anijain2305 commented on torch/_higher_order_ops/wrap.py:
```diff
@@ -110,24 +109,37 @@ def divide_kwargs(kwargs):
+                "Passing context_fn to torch.utils.checkpoint is currently not supported under torch.compile"
+            log.warning("""
+Detected that context_fn is passed to torch.utils.checkpoint under torch.compile.
+Please make sure the checkpointed region does not contain in-place ops (e.g. torch.relu_).
```
Maybe mention why do we restrict relu_ ops?

anijain2305 commented on test/dynamo/test_activation_checkpointing.py:
```diff
@@ -361,6 +413,298 @@ def fn(x, y):
+    @torch._dynamo.config.patch(
+        "_experimental_support_context_fn_in_torch_utils_checkpoint", True
+    )
+    @torch._functorch.config.patch("cse", True)
```
Do you need this?

anijain2305 commented on test/dynamo/test_activation_checkpointing.py:
```diff
@@ -361,6 +413,298 @@ def fn(x, y):
+        )
+        bw_compiler = functools.partial(
+            count_ops,
+            freq=4,
```
nit - Just a comment that we would expect 6 if not for SAC.

anijain2305 commented on test/dynamo/test_activation_checkpointing.py:
```diff
@@ -361,6 +413,298 @@ def fn(x, y):
+                torch.ops.aten.mm.default,
+            ]
+
+            def _custom_policy(mode, func, *args, **kwargs):
+                mm_count_key = f"{mode}_mm_count"
+                if mm_count_key not in meta:
+                    meta[mm_count_key] = 0
+                if func == torch.ops.aten.mm.default:
+                    meta[mm_count_key] += 1
+                # Saves output of all compute ops, except second mm
+                # (i.e. we will hint the partitioner to recompute second mm in backward pass)
+                return func in no_recompute_list and not (
+                    func == torch.ops.aten.mm.default and meta[mm_count_key] == 2
+                )
```
Nice! This is powerful.

anijain2305 commented on test/dynamo/test_activation_checkpointing.py:
```diff
@@ -361,6 +413,298 @@ def fn(x, y):
+        bw_compiler = functools.partial(
+            count_ops,
+            # The second mm may or may not be recomputed
+            # (since it's decided by the partitioner),
```
nit - I think its decided by the autograd dependencies, not really by the partitioner.

anijain2305 reviewed (commented):
Overall LGTM. Great work! Some nits here and there.

The only real comment is about unifying the `checkpoint` tracing between no context_fn and yes context_fn. 

Please also get a review from @soulitzer @albanD on the checkpoint related files.

anijain2305 reviewed (approved):
Chatted offline. Just add a comment about unification dependent on the inplace support in TorchDispatch models.

yf225 commented:
@pytorchbot merge

yf225 commented:
@pytorchbot label "topic: not user facing"

yf225 commented:
@pytorchbot merge

yf225 commented:
@pytorchbot rebase

yf225 commented:
@pytorchbot merge

yf225 commented:
@pytorchbot merge

yf225 commented:
@pytorchbot merge

yf225 commented:
@pytorchbot merge

yf225 commented:
@pytorchbot rebase

yf225 commented:
@pytorchbot merge

yf225 commented:
@pytorchbot rebase

yf225 committed:
torch.compile + selective activation checkpointing (#105489)

Summary:

NOTE: this PR is tagged "not user facing", because it's not ready to be announced externally yet.

This PR implements torch.compile + selective activation checkpoint (SAC) integration, by using `TagActivationCheckpoint` (same backend as torch.compile + full activation checkpoint integration).

TorchDispatchMode based implementation cannot support including inplace ops in the checkpointed region at the moment (the reason for this needs investigation), and there is also no way to ban them (because TorchDispatchMode now only sees "after-functionalization" ops, so can't detect if an op is in-place). Hence we hide torch.compile + SAC behind a flag (`torch._dynamo.config._experimental_support_context_fn_in_torch_utils_checkpoint`) and will only use it internally for cases that are known to not have in-place ops. This state won't last too long, because in-place op will at least be able to be detected after Brian's mode reordering and related functionalization changes.
So next steps after this PR:
1. Wait for Brian's mode reordering and related functionalization changes to land, and then try to enable the "inplace ops" unit test for torch.compile + selective activation checkpoint (if it doesn't work, investigate why).
2. Unify selective- and full-checkpoint under TorchDispatchMode based implementation.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng anijain2305 ipiszy


Differential Revision: D47497145

Pulled By: yf225

yf225 commented:
@pytorchbot merge -f "unrelated failures"
